File: Data Structure/Heap/LET_no295_Find Median from Data Stream.py
class MedianFinder:

    def __init__(self):
        self.first, self.second = [], []

    def addNum(self, num: int) -> None:
        heapq.heappush(self.second, -heapq.heappushpop(self.first, -num))
        if len(self.second) > len(self.first):
            heapq.heappush(self.first, -heapq.heappop(self.second))

    def findMedian(self) -> float:
        return -self.first[0] if len(self.first) > len(self.second) else (-self.first[0] + self.second[0]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: Data Structure/Heap/LET_no295_Find Median from Data Stream.cpp
class MedianFinder {
private: 
    priority_queue<int> first, second;
public:
    MedianFinder() {}
    
    void addNum(int num) {
        first.push(num);
        second.push(-first.top());
        first.pop();
        
        if (second.size() > first.size()) {
            first.push(-second.top());
            second.pop();
        }
    }
    
    double findMedian() {
        return first.size() > second.size() ? first.top() : (first.top() + (-second.top())) / 2.0;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */